### DISCUSSION TITLE
Regarding debugging kernel/firmware

### DISCUSSION THREAD

**Discussion by fazlamehrab** [cloud-hypervisor/cloud-hypervisor#4331] [Q&A] [answered]
Hi,
It is nice that the Cloud Hypervisor (CH) now supports gdb for remote debugging. It makes developers’ life easier. However, I was having some issues using this feature. Therefore, I was wondering if I am correctly using gdb.

I issue the following command in one terminal to launch the CH enabling remote debugging.

`sudo ~/cloud-hypervisor/target/debug/cloud-hypervisor --cpus boot=1 --memory size=512M --kernel ~/workloads/hypervisor-fw --disk path=~/workloads/osdisk.img path=~/workloads/cloudinit --net tap=,mac=12:34:56:78:90:01,ip=192.168.1.1,mask=255.255.255.0 --serial tty --console off --gdb path=/tmp/ch-gdb-sock`

In another terminal, I issue the following commands to connect to the gdb server and proceed with debugging.

```
sudo gdb -q
(gdb) target remote /tmp/ch-gdb-sock
Remote debugging using /tmp/ch-gdb-sock
warning: No executable has been specified and target does not support
determining executable automatically. Try using the “file” command.
0x0000000000111bfe in ?? ()
(gdb) file ~/rust-hypervisor-firmware/target/target/debug/hypervisor-fw
A program is being debugged already.
Are you sure you want to change the file? (y or n) y
Reading symbols from ~/rust-hypervisor-firmware/target/target/debug/hypervisor-fw...done.
(gdb) next
Remote connection closed
```

As the gdb says it cannot determine the executable, I specify the hypervisor-fw path with the gdb `file `command. However, when I issue the gdb `next` command, the CH instance throws the following error.

`cloud-hypervisor: 43.894165072s: <gdb> ERROR:vmm/src/gdb.rs:534 -- error occurred in GDB session: Target threw a fatal error: “Failed to resume the target: Vm(Debug(Resume(Resume(Invalid transition: InvalidStateTransition(Running, Running)))))”`

I get the same error for both `step` and `step i` commands.

If I am not doing it the wrong way, then my understanding is that the `resume (vmm/src/vm.rs:2592)` function is being invoked back-to-back twice when I issue one of these commands. Therefore, CH catches an invalid Running to Running VM state transition.

I don’t know the exact reason for this scenario. However, I figured that the following patch could be a temporary workaround.

``` 
diff --git a/vmm/src/vm.rs b/vmm/src/vm.rs
index 3b9f4a93..fede5a4b 100644
--- a/vmm/src/vm.rs
+++ b/vmm/src/vm.rs
@@ -2597,6 +2597,10 @@ impl Pausable for Vm {
.map_err(|e| MigratableError::Resume(anyhow!(“Could not get VM state: {}”, e)))?;
let new_state = VmState::Running;
+ if new_state == *state {
+ return Ok(());
+ }
+
state
.valid_transition(new_state)
.map_err(|e| MigratableError::Resume(anyhow!(“Invalid transition: {:?}”, e)))?;
```

Another problem with the `next` and `step` commands is that the execution stalls after issuing these commands. One workaround to this problem is to use the `step i` command instead. However, it may take a long time for the developers to reach the code they want to debug.

Fixing these issues would be a great help! However, there is a chance that I am doing the debugging improperly. In that case, please let me know the proper way to debug the kernel/firmware with CH.

Thanks,
Mehrab

**Comment by rbradford** [comment]
@retrage Perhaps you have some insight?

**Reply by rbradford** [reply]
Why do you need to load the binary in? Why not just start with the firmware specified on the command line with `--kernel`. The execution will stop ready for you to interact with it via gdb.

**Reply by fazlamehrab** [reply]
Yes, that would be the ideal scenario. The following snippet shows what it looks like if I do not load the binary. 

```
sudo gdb -q
(gdb) target remote /tmp/ch-gdb-sock
Remote debugging using /tmp/ch-gdb-sock
warning: No executable has been specified and target does not support
determining executable automatically.  Try using the "file" command.
0x0000000000111bfe in ?? ()
(gdb) next
Cannot find bounds of current function
(gdb) step
Cannot find bounds of current function
(gdb) step i
No symbol table is loaded.  Use the "file" command.
(gdb)
```

**Reply by rbradford** [reply]
The gdb instructions in the documentation only specify using raw addresses as breakpoints. I guess "stepi" isn't implemented or there is a bug.

**Reply by rbradford** [reply]
Using #4363 I was able to debug `hypervisor-fw` and successfully do `stepi`. To get symbols the trick was to start gdb with the binary: `gdb ~/src/rust-hypervisor-firmware/target/target/debug/hypervisor-fw -ex "target remote /tmp/ch-gdb-sock"`

**Reply by fazlamehrab** [reply]
This is great! Thanks a lot, Rob! 

Updating the gdb instructions accordingly in the documentation would really help. Any idea yet on why the `next` and `step` instructions are stalling the execution?

**Comment by michael2012z** [comment, marked as answer]
The problem in `next` and `step` commands were caused by some issues in VCPU state management, that could be some missing impact when GDB support was first introduced. The same problem was also seen when `continue` from a breakpoint.

The problem was fixed with PR https://github.com/cloud-hypervisor/cloud-hypervisor/pull/4363 . Using the latest version, GDB should function well.

**Reply by fazlamehrab** [reply]
I am probably doing something wrong then. Could you please tell me what that might be by looking at the following log?

1. Start Cloud Hypervisor with GDB:

```
$ git rev-parse HEAD
f4e219876726e5af03294952a399a19ad352c0f4
$ sudo ~/cloud-hypervisor/target/debug/cloud-hypervisor --cpus boot=1 --memory size=512M --kernel ~/rust-hypervisor-firmware/target/target/debug/hypervisor-fw --disk
path=~/workloads/osdisk.img  --net tap=,mac=12:34:56:78:90:01,ip=192.168.1.1,mask=255.255.255.0 --serial tty --console off --gdb path=/tmp/ch-gdb-sock
Setting up 4 GiB identity mapping
Page tables setup
```

2. Start GDB:

```
$ sudo gdb ~/rust-hypervisor-firmware/target/target/debug/hypervisor-fw -ex "target remote /tmp/ch-gdb-sock"
......(text ignored)
(gdb) hb main.rs:171
Hardware assisted breakpoint 1 at 0x115551: file src/main.rs, line 171.
(gdb) continue
Continuing.

Breakpoint 1, hypervisor_fw::main (info=...) at src/main.rs:172
172         log!("\nBooting with {}", info.name());
(gdb) next
(execution stalls here)
```

**Reply by rbradford** [reply]
Only `stepi` works for me. I don't think next/step would work. I assume because of the requirement to use hardware breakpoints (of which AFAICT only one works.)

**Reply by retrage** [reply]
@rbradford Thank you for fixing this bug. Yes, that's right. Because of the limitations of hardware breakpoints, it cannot support `next` and `step`.
